File: internal/auth/pathmatch.go
package auth

import "strings"

// pathMatcher implements the Ant-style two-level wildcard matching of
// Artifactory permission patterns (auth-model.md section 4, high
// confidence: o.a.a.util.PathMatcher over Spring AntPathMatcher, tokens
// trimmed of whitespace; '**' and '**/*' both mean "everything").
//
// Semantics implemented (kept deliberately minimal — M1 patterns are repo
// paths, not general Ant paths):
//
//   - pattern "" (or "**" or "**/*") matches every path;
//   - '*' matches any run of characters except '/';
//   - '**' matches any run including '/';
//   - a pattern ending in '/**' matches the directory itself and everything
//     under it ("ci-out/**" matches "ci-out" and "ci-out/a/b.bin");
//   - a prefix directory match (Ant matchStart) covers paths under a
//     pattern that names a directory — **but only when the path is itself a
//     folder** (trailing '/'): upstream enables matchStart exactly when
//     repoPath.isFolder() is true, so a *file* path never gains access
//     through a pattern that merely names one of its ancestor directories.
//     Convention (B-2 fix, mirrors repo.isFolderNode): a path ending in '/'
//     is a folder, everything else is a file. "ci-out" therefore matches
//     the folder "ci-out/" and anything below it, but NOT the file
//     "ci-out/a.bin" — use "ci-out/**" for that;
//   - matching is case-sensitive; segments are trimmed of surrounding
//     whitespace before comparison (AntPathMatcher token trimming).
type pathMatcher struct {
	pattern string
}

// match reports whether path (repo-relative, no leading '/') matches the
// pattern. A trailing '/' on path marks a folder and is the only form that
// can match through the directory-prefix rule.
func (m pathMatcher) match(path string) bool {
	pattern := m.pattern
	if pattern == "" || pattern == "**" || pattern == "**/*" {
		return true
	}
	path = strings.TrimSpace(path)
	if path == "" {
		return false
	}
	isFolder := strings.HasSuffix(path, "/")
	pSegs := splitSegments(pattern)
	tSegs := splitSegments(path)
	return segsMatch(pSegs, tSegs, isFolder)
}

// segsMatch matches pattern segments against path segments with backtracking
// on '**'. After all pattern segments are consumed:
//   - a path exactly consumed is a full match;
//   - leftover path segments are covered only when the path is a folder AND
//     the last pattern segment is a plain directory name (Ant matchStart
//     under its upstream isFolder() gate) — file paths must match fully.
func segsMatch(pSegs, tSegs []string, isFolder bool) bool {
	pi, ti := 0, 0
	lastPlain := false
	for pi < len(pSegs) {
		seg := pSegs[pi]
		switch {
		case seg == "**":
			// '**' at the end absorbs the whole rest (zero or more
			// segments); otherwise try every alignment of the remaining
			// pattern (classic Ant backtracking).
			rest := pSegs[pi+1:]
			for skip := ti; skip <= len(tSegs); skip++ {
				if segsMatch(rest, tSegs[skip:], isFolder) {
					return true
				}
			}
			return false
		case ti >= len(tSegs):
			return false
		case !segmentMatch(seg, tSegs[ti]):
			return false
		default:
			lastPlain = !strings.Contains(seg, "*")
			pi++
			ti++
		}
	}
	if ti == len(tSegs) {
		return true
	}
	// Pattern exhausted with path segments left: the directory-prefix rule
	// (matchStart) applies only to folder paths.
	return isFolder && lastPlain && pi > 0
}

// segmentMatch matches one path segment: '*' spans anything except '/'
// (already guaranteed by segmentation), all other characters compare
// literally.
func segmentMatch(pattern, segment string) bool {
	if !strings.Contains(pattern, "*") {
		return pattern == segment
	}
	return wildcardEqual(pattern, segment)
}

// wildcardEqual compares a segment containing '*' with a concrete segment.
func wildcardEqual(pattern, s string) bool {
	parts := strings.Split(pattern, "*")
	if len(parts) == 1 {
		return pattern == s
	}
	// Must start with the first literal and end with the last.
	if !strings.HasPrefix(s, parts[0]) {
		return false
	}
	s = s[len(parts[0]):]
	last := parts[len(parts)-1]
	if !strings.HasSuffix(s, last) {
		return false
	}
	s = s[:len(s)-len(last)]
	// Middle literals must appear in order.
	for _, mid := range parts[1 : len(parts)-1] {
		if mid == "" {
			continue
		}
		idx := strings.Index(s, mid)
		if idx < 0 {
			return false
		}
		s = s[idx+len(mid):]
	}
	return true
}

// splitSegments splits a pattern/path on '/' and trims whitespace from each
// segment (Ant token trimming), dropping empties. A trailing '/' therefore
// leaves no marker segment; folder-ness is tracked separately by the caller.
func splitSegments(s string) []string {
	raw := strings.Split(s, "/")
	segs := make([]string, 0, len(raw))
	for _, seg := range raw {
		seg = strings.TrimSpace(seg)
		if seg != "" {
			segs = append(segs, seg)
		}
	}
	return segs
}

// matchesAny reports whether path matches any of the patterns (an empty
// pattern list contributes nothing; callers treat "no include patterns" as
// "include everything" separately, per auth-model.md section 4).
func matchesAny(patterns []string, path string) bool {
	for _, p := range patterns {
		if (pathMatcher{pattern: p}).match(path) {
			return true
		}
	}
	return false
}

File: web/src/pages/security/pathmatch.ts
// 权限 pattern 判定（console-ux §4.9[2]/§9-R8：模式测试器与服务端
// internal/auth/pathmatch.go 必须同源）。
//
// 本文件是 pathmatch.go 的逐语义 TypeScript 移植：
//   - pattern "" / "**" / "**/*" 匹配一切路径；
//   - '*' 跨段内任意字符（不跨 '/'）；'**' 跨任意段（Ant 回溯）；
//   - 以 '/**' 结尾的 pattern 匹配目录自身与其下一切；
//   - 目录前缀规则（Ant matchStart）仅在路径本身是目录（尾 '/'）时生效
//     ——文件路径必须整路径全匹配（B-2 修复语义）；
//   - 大小写敏感；段两侧空白在比较前裁剪（Ant token trimming）。
//
// 同源性由两道闸保证（无后端端点，ux R8 定案方案）：
//   1. pathmatch.fixtures.ts 由 scripts/gen-pathmatch-fixtures.mjs 从
//      internal/auth/pathmatch_test.go 的 table-driven 用例生成（漂移即红）；
//   2. e2e/pathmatch-parity.spec.ts 对全部 fixtures 断言本实现逐条一致。
//
// evaluatePath 在匹配之上叠加 authorizer.targetCovers 的组合规则：
// 空 includes = 匹配全部；任一 exclude 命中一票否决（exclude 优先）。

export interface PatternVerdict {
  pattern: string
  hit: boolean
}

export interface PathEvaluation {
  /** 每个 include pattern 的命中明细（§4.9「逐条命中明细」） */
  includes: PatternVerdict[]
  /** 每个 exclude pattern 的命中明细 */
  excludes: PatternVerdict[]
  /** includes 为空时的「匹配全部」语义由 includesEmpty 表达 */
  includesEmpty: boolean
  /** 命中的第一条 include（includesEmpty 时为 null） */
  includedBy: string | null
  /** 命中的第一条 exclude（一票否决项） */
  excludedBy: string | null
  /** 最终判定（exclude 优先） */
  match: boolean
}

function splitSegments(s: string): string[] {
  const segs: string[] = []
  for (const raw of s.split('/')) {
    const seg = raw.trim()
    if (seg !== '') segs.push(seg)
  }
  return segs
}

/** 段内 '*' 通配：首尾字面量锚定 + 中间字面量按序出现（Ant 段语义） */
function wildcardEqual(pattern: string, s: string): boolean {
  const parts = pattern.split('*')
  if (parts.length === 1) return pattern === s
  if (!s.startsWith(parts[0])) return false
  s = s.slice(parts[0].length)
  const last = parts[parts.length - 1]
  if (!s.endsWith(last)) return false
  s = s.slice(0, s.length - last.length)
  for (const mid of parts.slice(1, -1)) {
    if (mid === '') continue
    const idx = s.indexOf(mid)
    if (idx < 0) return false
    s = s.slice(idx + mid.length)
  }
  return true
}

function segmentMatch(pattern: string, segment: string): boolean {
  if (!pattern.includes('*')) return pattern === segment
  return wildcardEqual(pattern, segment)
}

/** '**' 回溯匹配（与 pathmatch.go segsMatch 逐行对应） */
function segsMatch(pSegs: string[], tSegs: string[], isFolder: boolean): boolean {
  let pi = 0
  let ti = 0
  let lastPlain = false
  while (pi < pSegs.length) {
    const seg = pSegs[pi]
    if (seg === '**') {
      const rest = pSegs.slice(pi + 1)
      for (let skip = ti; skip <= tSegs.length; skip++) {
        if (segsMatch(rest, tSegs.slice(skip), isFolder)) return true
      }
      return false
    }
    if (ti >= tSegs.length) return false
    if (!segmentMatch(seg, tSegs[ti])) return false
    lastPlain = !seg.includes('*')
    pi++
    ti++
  }
  if (ti === tSegs.length) return true
  // pattern 耗尽而路径还有余段：目录前缀规则只对目录路径生效
  return isFolder && lastPlain && pi > 0
}

/** 单 pattern 匹配（pathMatcher.match 移植；path 为 repo 相对路径，无前导 '/'） */
export function pathMatch(pattern: string, path: string): boolean {
  if (pattern === '' || pattern === '**' || pattern === '**/*') return true
  path = path.trim()
  if (path === '') return false
  const isFolder = path.endsWith('/')
  return segsMatch(splitSegments(pattern), splitSegments(path), isFolder)
}

export function matchesAny(patterns: string[], path: string): boolean {
  return patterns.some((p) => pathMatch(p, path))
}

/**
 * 模式测试器判定（§4.9 线框）：
 * 空 includes = 匹配全部路径（auth-model §4 / targetCovers 同款）；
 * 任一 exclude 命中即否决（exclude 优先）。
 */
export function evaluatePath(includes: string[], excludes: string[], path: string): PathEvaluation {
  const inc = includes.map((pattern) => ({ pattern, hit: pathMatch(pattern, path) }))
  const exc = excludes.map((pattern) => ({ pattern, hit: pathMatch(pattern, path) }))
  const includesEmpty = includes.length === 0
  const hitInc = inc.find((v) => v.hit) ?? null
  const hitExc = exc.find((v) => v.hit) ?? null
  const included = includesEmpty || hitInc !== null
  const match = included && hitExc === null
  return {
    includes: inc,
    excludes: exc,
    includesEmpty,
    includedBy: hitInc?.pattern ?? null,
    excludedBy: hitExc?.pattern ?? null,
    match,
  }
}
